Feature/simcom

KraPete opened the pull request:
Патчи для поддержки SIM A7600E-H:
1. Поправил разбор ответа на `AT+CGDCONT?`. Старый регэксп расчитывал, что есть минимум 4 параметра в ответе. По стандарту можно меньше. Наш модем присылает 3. В результате не работала процедура проверки того, что контекст создан.
2. Убрал `-Werror`. Вылезла куча мест, где компилятор ругается на преобразование выравнивания (`Wcast-align`). Возможно, надо как-то иначе собирать, но я не понял как. `dpkg-buildpackage -us -uc` приводит к этим ошибкам, без них всё собирается и работает.
3. MM запрашивает у модема типы IP адресов, которые тот поддерживает. Модем возвращает ошибку, если не зареган в сети. К сожалению, эта процедура выполняется один раз за всё время, и результат потом кэшируется. Получается так, что MM не видит ни одного типа IP адресов. При создании поключения NetworkManager офигевает от такого расклада и падает в сегфолт. Отдаю для SIM A7600E-H все возможные типы без запроса к модему.

KraPete committed:
Fix CGDCONT read response parsing

SIM A760E-H sends only 3 parameters in CGDCONT read response (+CGDCONT: 1,"IP","nate.sktelecom.com").

KraPete committed:
Disable -Werror

KraPete committed:
Set IP families for SIM A7600E-H

Supported IP families request is performed only once and SIM A7600E-H can return ERROR if internal logic is not ready. MM can't define IP families and send them to NetworkManager.       It will lead to NM crash. So set IP families based on model name without querying a modem.

evgeny-boger reviewed (changes requested):
нам же эти два патча надо отправлять в апстрим. 

Поэтому нужно:
1) от первого оторвать коммит в наш debian/changelog
2) перенести из описания PR подробное описание проблемы в commit message
3) отправить это как патчи в апстрим

KraPete commented:
> нам же эти два патча надо отправлять в апстрим.
> 
> Поэтому нужно:
> 
>     1. от первого оторвать коммит в наш debian/changelog
> 
>     2. перенести из описания PR подробное описание проблемы в commit message
> 
>     3. отправить это как патчи в апстрим

Не уверен, что всё могут принять. Как минимум последний патч - скорее костыль. Думаю, надо менять логику запроса списка поддерживаемых типов IP, т.е. делать это где-то в другом месте или иначе. Это субъективно переделывание внутренней логики MM. Оно нам надо? Плюс в bullseye сейчас версии MM и NM не из upstream.

evgeny-boger commented:
ну конечно надо, потому что иначе тебе придётся это раз в полгода натягивать на новый modemanager, а это быстро надоест. Можно отправить как есть, заодно кто-нибудь расскажет как это сделать правильно.

KraPete committed:
Simtech SIM switch imlementation

KraPete commented:
@evgeny-boger @webconn Последни коммитом добавил переключение симок чрезе стандартные команды MM
![изображение](https://user-images.githubusercontent.com/86825564/183088862-bc4e656e-0b52-4e13-8187-b6acaadd74fd.png)

Код в части работы с gpio - полное г. Основной вопрос в том, как мы будем передавать в MM информацию, что у нас есть 2 слота, и какой gpio за это всё отвечает. Сейчас MM умеет читать настройки из udev, т.е. правилом добавляются параметры в устройство, и MM их вычитывает. Предлагаю сделать аналогично. Можно при старте генерить правило, с нужным номером gpio.

KraPete committed:
Sim switching using libgpiod

KraPete commented:
Сделал переключение симки через libgpiod. Добавил параметр `ID_MM_SIM_SWITCH_GPIO_LABEL` с меткой gpio, чтобы сигнализировать MM о том, как переключать симки.
Пример правила для A7600E-H, которое объявляет такое gpio:
```
ATTRS{idVendor}=="1e0e", ATTRS{idProduct}=="9011", ENV{ID_MM_SIM_SWITCH_GPIO_LABEL}="SIM_SELECT"
```

KraPete committed:
Check GPIO avalibility before switching SIM

KraPete committed:
Increase CRSM read timeouts

KraPete committed:
Handle URC

KraPete committed:
Ignore NO CARRIER URC during SIM switching

A7600E-H can send NO CARRIER URC after AT+CFUN=0 but before sending OK. ModemManager will treat the situation as error. So enable NO CARRIER filtering during AT+CFUN=0

webconn commented on plugins/simtech/mm-broadband-modem-simtech.c:
```diff
@@ -1060,6 +1139,468 @@ load_current_modes (MMIfaceModem        *self,
+}
+
+static MMBearerIpFamily
+load_supported_ip_families_finish (MMIfaceModem *self,
```
Этот метод не по умолчанию разве? Я так понимаю, он скопипащен

KraPete replied:
Да. Я посмотрел как остальные плагины сделаны, там тоже копируют, вот и перенёс.

webconn replied:
Вот этот метод вроде никто не копировал... Его по наследству никак не получить?

KraPete replied:
Обычно делают пару `запрос-finish`. Можно из `iface_modem_parent` взять

KraPete replied:
Опасность в том, что поменяется логика метода в родителе, и тут мы ничего не узнаем.

webconn replied:
Ну тут же finish полностью идентичен, вряд ли есть смысл копипастить

webconn replied:
> Опасность в том, что поменяется логика метода в родителе, и тут мы ничего не узнаем.

Если от смены логики в родителе что-то будет зависеть, всё и так наверняка на нашем коде сломается, потому что только мы его и поддерживаем

KraPete replied:
Я имел ввиду логику `запрос-finish`. Сигнатура запроса и `finish` задана в интерфейсе, она не изменится. `finish` интерпретирует промежуточный результат запроса и выдаёт что-то вызвавшему. И вот этот промежуточный результат - особенность реализации. Он может измениться, и finish у родителя тоже.

webconn replied:
Ладно, давай оставим это как есть.

webconn commented on plugins/simtech/mm-broadband-modem-simtech.c:
```diff
@@ -66,8 +68,64 @@ struct _MMBroadbandModemSimtechPrivate {
+} CPinResult;
+
+static CPinResult unlock_results[] = {
+    /* Longer entries first so we catch the correct one with strcmp() */
```
Они же сравниваются не `strcmp()`, а `g_str_has_prefix()`, вот тогда становится понятно, почему тут это важно

KraPete replied:
Убрал вообще

webconn commented on plugins/simtech/mm-broadband-modem-simtech.c:
```diff
@@ -1060,6 +1139,468 @@ load_current_modes (MMIfaceModem        *self,
+}
+
+/*****************************************************************************/
+/* Load sim slots (Modem interface) */
```
Комментарий несколько раз повторяется

KraPete replied:
Исправил

webconn commented on src/mm-base-sim.c:
```diff
@@ -2115,7 +2115,7 @@ load_operator_identifier (MMBaseSim *self,
         self->priv->modem,
         "+CRSM=176,28589,0,0,4",
-        10,
+        20,
```
Так критичен оказался этот таймаут?

KraPete replied:
Yes! У меня локально буквально секунды не хватало. В доке у симкома ничего не написано про таймаут, так что увеличил в 2 раза с запасом

webconn replied:
Жуть. Меня смущает, что это в общем коде тут происходит, и об этом никак не догадаешься. Хочу комментарий в этом месте, и запись в changelog, что мы это сделали осознанно

KraPete replied:
Можно утащить это всё в плагин. Будет гора копипасты ради одной константы.

webconn replied:
Не, давай не будем, просто подпишем, что это место хрупкое

webconn replied:
Всё ещё актуально

KraPete replied:
Я по коду посмотрел, в других местах при вызове этой команды уже стоит 20 секунд таймаута. Просто тут забыли. Может, не писать комменты? А то придётся и в остальных местах добавлять.

webconn replied:
А в других-то зачем? Смысл этих комментариев в том, чтобы обозначить наши собственные изменения. Может пригодиться, если в апстриме тоже поменяют, будет чуть проще дифф читать

KraPete replied:
Написал

webconn commented:
И как в итоге с патчами? Проводить ревью у нас в этом плане смысла не очень много, лучше сразу забросить это в сообщество. Если с нашей стороны сейчас всё работает, надо мержить, а потом по мере принятия патчей переносить то, что не примут никогда

KraPete commented:
> И как в итоге с патчами? Проводить ревью у нас в этом плане смысла не очень много, лучше сразу забросить это в сообщество. Если с нашей стороны сейчас всё работает, надо мержить, а потом по мере принятия патчей переносить то, что не примут никогда

Я бы довёл всё до нашего testing, поправил всё, что вылезет на широкой публике, и потом начал патчи слать.

KraPete committed:
Fix comments

KraPete committed:
More comments

KraPete commented:
Rebase https://github.com/wirenboard/modemmanager/pull/3
